text/0000-match-ergonomics.md

```diff
@@ -0,0 +1,455 @@
+bindings as `ref` or `ref mut` bindings.
+
+Example:
+
+```rust
+let x = Some(3);
+let y: &Option<i32> = &x;
+match y {
+  Some(a) => {
+    // `y` is dereferenced, and `a` is bound like `ref a`.
+  }
+  None => {}
+}
+```
+
+Note that this RFC applies to all instances of pattern-matching, not just
+`match` expressions:
+
+```rust
+struct Foo(i32);
+
+let foo = Foo(6);
+let foo_ref = &foo;
+// `foo_ref` is dereferenced, and `x` is bound like `ref x`.
+let Foo(x) = foo_ref;
+```
+
+
+## Definitions
+
+A reference pattern is any pattern which can match a reference without
+coercion. Reference patterns include bindings, wildcards (`_`),
+`const`s of reference types, and patterns beginning with `&` or `&mut`. All
+other patterns are _non-reference patterns_.
+
+_Default binding mode_: this mode, either `move`, `ref`, or `ref mut`, is used
+to determine how to bind new pattern variables.
+When the compiler sees a variable binding not explicitly marked
+`ref`, `ref mut`, or `mut`, it uses the _default binding mode_
+to determine how the variable should be bound.
+Currently, the _default binding mode_ is always `move`.
+Under this RFC, matching a reference with a _non-reference pattern_, would shift
+the default binding mode to `ref` or `ref mut`.
+
+## Binding mode rules
+
+The _default binding mode_ starts out as `move`. When matching a pattern, the
+compiler starts from the outside of the pattern and works inwards.
+Each time a reference is matched using a _non-reference pattern_,
+it will automatically dereference the value and update the default binding mode:
+
+1. If the reference encountered is `&Pat`, set the default binding mode to `ref`.
+2. If the reference encountered is `&mut Pat`: if the current default
+binding mode is `ref`, it should remain `ref`. Otherwise, set the current binding
+mode to `ref mut`.
+
+```
+                        Start                                
+                          |                                  
+                          v                                  
+                +-----------------------+                     
+                | Default Binding Mode: |                     
+                |        move           |                     
+                +-----------------------+                     
+               /                        \                     
+Encountered   /                          \  Encountered       
+  &mut Pat   /                            \     &Pat
+            v                              v                  
++-----------------------+        +-----------------------+    
+| Default Binding Mode: |        | Default Binding Mode: |    
+|        ref mut        |        |        ref            |    
++-----------------------+        +-----------------------+    
+                          ----->                              
+                        Encountered                           
+                            &Pat                              
```

Comment:
The "Encountered &Pat" labels in this flowchart are confusing for me. As I understand the RFC, it means "matched type is `&Pat`, but encountered pattern of type `Pat`", which is something different than this flowchart is suggesting.